### PR TITLE
[hmac] Implementation of wider digest & configurable key length

### PR DESCRIPTION
**Pull request by gdessouky**
This brings the implementation of the wider digest size and configurable key length for HMAC (issue #20968).  This should allow the HMAC IP to compute SHA-2 and keyed HMAC using either SHA-256, SHA-384 or SHA-512. This also allows configuring the key length to one of the supported 128-bit/256-bit/384-bit/512-bit/1024-bit.  If a key length larger than the block size (which is twice the digest length) of a configured mode is picked, this gets capped to the block size, i.e., if 1024-bit key is selected with SHA-256, this will get capped at 512-bit to simplify the HW implementation. When a larger key length is required, the SW is expected to manage this by hashing the key and then supplying that hashed key to the HW. 

SW usage requires additionally configuring the digest size and key length required - DIFs will need to be modified accordingly. Unsupported key lengths and digest sizes are not handled or tested at the moment; most conservative/correct approach is to not proceed with the operation and flag an error, and leave the digest CSRs untouched.

This PR also brings changes to the DV environment to provide preliminary support for the extended features and to pass the basic smoke test. I haven't modified other tests yet, and definitely don't see the full regression tests running without extending them all appropriately.

Note that this will need to be merged together with the save & restore implementation in PR  #21307 and tested again to pass basic smoke.

Follow-up PRs needed:
- [ ] update documentation and programmer's guide
- [ ] update DIFs
- [x] update HMAC version number 
- [ ] handle unsupported digest size and key length configuration: flag an error and leave digest CSRs retaining previous digest. These are not dealt with at the moment in the RTL and DV.
- [ ] 1 or 2 DV checks/assertions trigger issues which I have commented them for now (labelled with TODO); I believe they are FIFO status issues that have come up because of the integration of the HMAC core with the 32-bit wrapper for the wider SHA-384/512 engine, and will need DV adaptations. 
- [ ] extend DV support for the rest of the existing tests
- [ ] extend testplan to test new cases; e.g., supply unsupported digest size or key length configuration, configuration changes during runtime, etc.

### CODE REVIEW

**Comment by gdessouky** [comment]
After aligning with @andreaskurth, it's more ideal if we would merge this only after we confirm that original functionality (mode 256) hasn't been broken; this requires extending support to the full regression tests to pass for SHA-2 256.

File: hw/ip/hmac/rtl/hmac_core.sv
Comment (vogelpi) [inline hw/ip/hmac/rtl/hmac_core.sv:245]: This can be done using a `unique case` (I don't remember the exact syntax offhand but it will be more readable)
```sv
always comb begin
  unique_case (digest_size):
    SHA2_256: txcnt_eq_blksz = (txcount[BlockSizeBitsSHA256:0] == BlockSizeBSBSHA256);
    SHA2_384,
    SHA2_512: txcnt_eq_blksz = (txcount[BlockSizeBitsSHA512:0] == BlockSizeBSBSHA512);
    default: '0;
end
```

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [inline hw/ip/hmac/data/hmac.hjson:9]: As we anyway change parts of this file now, I recommend changing the touched lines to 1 sentence per line. It will ease future diffs.

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [inline hw/ip/hmac/data/hmac.hjson:237]: Nice, this seems familiar to me :-)
Here we should also specify what happens if that value is set once a decision has been taken.

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [inline hw/ip/hmac/data/hmac.hjson:10]: A note on consistency: in this file, currently both SHA-256 and SHA2-256 etc. is used, I suggest choosing one of the two and then using it consistently.

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (vogelpi) [inline hw/ip/prim/rtl/prim_sha2_32.sv:73]: This doesn't look right.

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [inline hw/ip/hmac/data/hmac.hjson:118]: I think the feature names are all caps -> HMAC.KEY_LENGTH.KEY_128

**Comment by gdessouky** [comment]
I've force-pushed after rebasing and integrating with the save & restore changes in https://github.com/lowRISC/opentitan/pull/21307, and getting all block-level tests pass at 100% still. All do now, except `hmac_stress_all_with_rand_reset` at 7.50%, which is a V3 test, and I need to look more closely into it. I think this can wait after M2.

Thanks @vogelpi for your RTL review! I'll be addressing your comments in the next forced push, but want to try and get a first RTL implementation merged in. 

I'll need to update the DIFs to pass the TLTs exercised in CI.

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv]: Corret, and agree, this change is already underway now because it also reduces the change to the drivers and DIFs.

File: hw/ip/hmac/rtl/hmac_core.sv
Comment (andreaskurth) [inline hw/ip/hmac/rtl/hmac_core.sv:56]: No longer needed? suggest to remove

File: hw/ip/hmac/rtl/hmac_core.sv
Comment (andreaskurth) [reply hw/ip/hmac/rtl/hmac_core.sv:182]: I agree with @vogelpi, though I think we can resolve this in a follow-up PR to not gate this one longer than absolutely needed

File: hw/ip/prim/rtl/prim_sha2.sv
Comment (andreaskurth) [inline hw/ip/prim/rtl/prim_sha2.sv:147]: I don't understand these whitespace changes; were they accidental?

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (andreaskurth) [inline hw/ip/prim/rtl/prim_sha2_32.sv:81]: Why do we no longer need this code? And can we remove it instead of just commenting it out?

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (andreaskurth) [inline hw/ip/prim/rtl/prim_sha2_32.sv:191]: Nit: keep comment aligned, please

**Review by andreaskurth — COMMENTED** [review]
Thanks a lot for this work, @gdessouky :+1:

I have some suggestions (mostly small ones) on the RTL. I'll review the DV side in a second pass.

File: hw/ip/hmac/data/hmac.hjson
Comment (andreaskurth) [inline hw/ip/hmac/data/hmac.hjson:498]: Do we need to leave so much space in the register map between CSRs and message FIFO? Wouldn't a skip of `0x1000` suffice (reserving 4 KiB for the CSRs)

File: hw/ip/hmac/data/hmac.hjson
Comment (andreaskurth) [reply hw/ip/hmac/data/hmac.hjson:478]: Just to put this into perspective, hashing 2^64 bit at a rate of 32 bit per cycle (the most the current interface allows) at 100 MHz takes more than 182 years. I really think we don't need the message length to be wider than 64 bit :-)

**Comment by andreaskurth** [comment]
We're working on fixing the failing tests. Essentially `make -C hw` is missing and some patches to SW using HMAC. Once this is concluded, what I think should minimally be changed before we can merge this is:
- [ ] Change the `MSG_LENGTH` CSRs back to 2x32 bit
- [ ] Decrease the address skip before `MSG_FIFO`.

All other opens, I think, can be addressed in follow-up PRs, and we should create issues to track them (after we have merged this PR).

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv:267]: Thanks, also agree, I haven't had the time to change it yet, and for now wanted to map either invalid digest size or key length to what was already existing, i.e. SHA-2 256 and 256-bit key. I think it's best/most conservative to map either unsupported digest modes/key lengths to SHA2_None /Key_None and not run altogether and reflect that in an error status. I'll track these fixes in a couple of issues I'll be spinning out after this PR lands.

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv:451]: Yes, for the last/default case when `digest_size = SHA2_None`, otherwise it is a latch.

File: hw/ip/hmac/data/hmac.hjson
Comment (gdessouky) [reply hw/ip/hmac/data/hmac.hjson]: Exactly, once I got to the SW side, it became obvious it's better to keep the SHA-2 256 digest mapping unchanged to DIGEST0 - DIGEST7, and have taken care of that.

File: hw/ip/hmac/rtl/hmac.sv
Comment (msfschaffner) [inline hw/ip/hmac/rtl/hmac.sv:232]: if this is still creating latches in synthesis, I would investigate whether declaring`digest` as a `hwext` register could cause issues here. i.e., some way `.q` is dependent on `.d`, which could be a combinational path in that case due to the `hwext` declaration...

Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv:232]: not anymore after the last push, but the logic here can use some cleanup/reorg a bit esp with the defaults assignment but after we get past a successful a CI run.

**Review by andreaskurth — APPROVED** [review]
Thanks @gdessouky for implementing our feedback and getting this green! The remaining opens can be resolved in follow-up PRs, so LGTM --> merging!

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (cdgori) [reply hw/ip/prim/rtl/prim_sha2_32.sv:81]: I haven't checked forward yet to see if it was resolved later, but this seems strange to me as well, and might need a TODO / tracking issue.

**Review by cdgori — COMMENTED** [review]
LGTM modulo the one possible issue in prim_sha2_32.sv

File: hw/ip/hmac/rtl/hmac.sv
Comment (andreaskurth) [reply hw/ip/hmac/rtl/hmac.sv:215]: Resolved in #22580

Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv:215]: @cdgori my first intuition was to also clear digests, but we will be retaining the previous digest in CSRs for the sake of SW and in case of invalid configurations etc.  This is concerning with context switching, but as long as `sha_en` is de-asserted, context and digests get cleared, and so digests will not get leaked to another SW thread.

**Pull request merged**
